# Digit Factorials

## Problem 34

145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145. As 1! = 1 and 2! = 2 are not sums they are not included.

Find the sum of all numbers which are equal to the sum of the factorial of their digits.

### Solution
The difficulty of this task is to find the upper bound for checking if numbers are curious.

Let's first create a function calculating sum of factorials of digits of any number:

In [1]:
# Function calculating factorial
def factorial(n):
    if n < 0:
        raise Exception("Value must be non-negative")
    factorial_n = 1
    for i in range(1, n+1):
        factorial_n *= i
    return factorial_n
'''
Alternatively, use Python library:
import math
math.factorial(n)
'''

# Hash table with a factorial of digits 1-9 - to save computing time
digit_factorials = {i: factorial(i) for i in range(10)}

# Function calculating sum of digit factorials
def factorial_sum(n):
    return sum(digit_factorials[int(digit)] for digit in str(n))

Now, let's check maximum value of sums of digit factorials for numbers with 2, 3, 4, ..., 10 digits.

In [3]:
# Generate max number for each length (1 to 10 digits)
max_digits_nums = [(10 ** k) - 1 for k in range(1, 11)]  # 9, 99, 999, ..., 9999999999

# Find sum of digit factorials for max numbers with 1, 2, 3, ..., 10 digits
max_factorial_sums = {len(str(n)): factorial_sum(n) for n in max_digits_nums}
print("Max sums of digit factorials for max 1, 2, 3..., 10-digit numbers:")
print(max_factorial_sums)

# Calculate number of digits for each factorial sum and compare lengths
lengths_factorial_sums = {n: len(str(factorial_sum)) for n, factorial_sum in max_factorial_sums.items()}
print("\nTherefore, max lengths of sums of digit factorials for 1, 2, 3..., 10-digit numbers are:")
print(lengths_factorial_sums)

Max sums of digit factorials for max 1, 2, 3..., 10-digit numbers:
{1: 362880, 2: 725760, 3: 1088640, 4: 1451520, 5: 1814400, 6: 2177280, 7: 2540160, 8: 2903040, 9: 3265920, 10: 3628800}

Therefore, max lengths of sums of digit factorials for 1, 2, 3..., 10-digit numbers are:
{1: 6, 2: 6, 3: 7, 4: 7, 5: 7, 6: 7, 7: 7, 8: 7, 9: 7, 10: 7}


Since for > 8-digit numbers we can only construct sum of digit factorials which is max 7-digit long, the upper bound for finding curious numbers is the max factorial sum for number with 7-digits:

In [4]:
upper_bound = max_factorial_sums[7]
print(upper_bound)

2540160


Having established the upper bound = 2540160, let's the sum of all curious numbers.

In [5]:
# Function for checking computing time
import time

# Function to calculate execution time
def compute_time(func):
    start_time = time.time()
    result = func()
    end_time = time.time()
    total_time = end_time - start_time
    return result, total_time

In [6]:
# Your main function that finds curious numbers
def find_curious_numbers():
    curious_nums = {factorial_sum(n) for n in range(3, upper_bound + 1) if factorial_sum(n) == n}
    all_curious_nums_sum = sum(curious_nums)
    return curious_nums, all_curious_nums_sum

In [7]:
# Call compute_time to measure the execution of find_curious_numbers
result, elapsed_time = compute_time(find_curious_numbers)

# Unpack the results
curious_nums, all_curious_nums_sum = result

# Print the results
print(f"The set of all curious numbers is: {curious_nums} and its sum is {all_curious_nums_sum}.")
print(f"Computation time: {elapsed_time:.2f} s")

The set of all curious numbers is: {145, 40585} and its sum is 40730.
Computation time: 3.58 s
